File: Semester 4/Programming and Projection Enironments/mpp-proiect-csharp/Proiect_MPP/Proiect_MPP/service/main/MainService.cs
﻿using Proiect_MPP.domain;
using Proiect_MPP.repository.flights;
using Proiect_MPP.repository.airports;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Proiect_MPP.service.main
{
    public class MainService : IMainService
    {
        private readonly FlightRepository flightRepository;
        private readonly AirportRepository airportRepository;

        public MainService(FlightRepository flightRepository, AirportRepository airportRepository)
        {
            this.flightRepository = flightRepository;
            this.airportRepository = airportRepository;
        }

        public void updateFlight(Flight flight, int id)
        {
            flightRepository.update(flight, id);
        }

        public Airport? findAirportById(int id)
        {
            return airportRepository.findById(id);
        }

        public Flight? findFlightById(int id)
        {
            return flightRepository.findById(id);
        }

        public List<Flight> getAllAvailableFlights()
        {
            return flightRepository.getAvailable();
        }

        public List<Flight> findByDestinationDate(string destination, DateTime Date)
        {
            List<Airport> airports = airportRepository.getAirportsInCity(destination);
            List<Flight> flightList = new List<Flight>();
            List<Flight> flights;
            DateTime departureDate;
            foreach (Airport airport in airports) 
            {
                flights = flightRepository.getByDestinationAirport(airport.ID);
                if (null == flights)
                {
                    continue;
                }

                foreach (Flight flight in flights)
                {
                    departureDate = flight.DepartureDate;
                    if (departureDate.Year != Date.Year ||
                        departureDate.Month != Date.Month || 
                        departureDate.Day != Date.Day ||
                        flight.FreeSeats == 0)
                    {
                        continue;
                    }
                    flightList.Add(flight);
                }
            }
            return flightList;
        }

    }
}

File: Semester 4/Programming and Projection Enironments/mpp-proiect-java/Proiect_MPP/Thrift/src/main/java/root/proiect_mpp/service/main/MainService.java
package root.proiect_mpp.service.main;

import root.proiect_mpp.domain.Airport;
import root.proiect_mpp.domain.Flight;
import root.proiect_mpp.repositories.airports.AirportRepository;
import root.proiect_mpp.repositories.flights.FlightRepository;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class MainService implements IMainService {

    private final FlightRepository flightRepository;
    private final AirportRepository airportRepository;

    public MainService(FlightRepository flightRepository, AirportRepository airportRepository) {
        this.flightRepository = flightRepository;
        this.airportRepository = airportRepository;
    }

    @Override
    public void updateFlight(Flight elem, int id) {
        flightRepository.update(elem, id);
    }

    @Override
    public Airport findAirportById(int id) {
        return airportRepository.findById(id);
    }

    @Override
    public Flight findFlightById(int id) {
        return flightRepository.findById(id);
    }

    @Override
    public Collection<Flight> getAllAvailableFlights() {
        return flightRepository.getAvailable();
    }

    @Override
    public Collection<Flight> findByDestinationDate(String destination, LocalDate date) {
        Collection<Airport> airports = airportRepository.getAirportsInCity(destination);
        List<Flight> flightList = new ArrayList<>();
        Collection<Flight> flights;
        LocalDate departureDate;
        for (Airport airport : airports) {
            flights = flightRepository.getByDestinationAirport(airport.getId());
            if (null == flights) {
                continue;
            }
            for (Flight flight : flights) {
                departureDate = flight.getDepartureDate();
                if (departureDate.getYear() != date.getYear() ||
                        departureDate.getMonthValue() != date.getMonthValue() ||
                        departureDate.getDayOfMonth() != date.getDayOfMonth() ||
                        flight.getFreeSeats() == 0) {
                    continue;
                }
                flightList.add(flight);
            }
        }
        return flightList;
    }
}
